xchange-cexio/src/main/java/info/bitrich/xchangestream/cexio/CexioStreamingRawService.java

```diff
@@ -23,44 +30,139 @@
     public static final String PONG = "pong";
     public static final String ORDER = "order";
     public static final String TRANSACTION = "tx";
+    public static final String ORDERBOOK = "order-book-subscribe";
+    public static final String ORDERBOOK_UPDATE = "md_update";
 
     private String apiKey;
     private String apiSecret;
-
+    private boolean isAuthenticated = false;
+    private final Lock authLock = new ReentrantLock();
+    private final Condition authCond = authLock.newCondition();
+    private final boolean isTest;
+    
     private PublishSubject<Order> subjectOrder = PublishSubject.create();
     private PublishSubject<CexioWebSocketTransaction> subjectTransaction = PublishSubject.create();
-
-    public CexioStreamingRawService(String apiUrl) {
+    
+    public CexioStreamingRawService(String apiUrl, boolean isTest) {
         super(apiUrl, Integer.MAX_VALUE);
+        this.isTest = isTest;
     }
-
+    
+    public void waitForAuthIfNeeded() {
+    	authLock.lock();
+    	try
+    	{
+    		while (!isAuthenticated && !isTest)
+    		{
+    			authCond.await();
+    		}
+    	} catch (InterruptedException e) {
+			// TODO Auto-generated catch block
+			e.printStackTrace();
+		}
+    	finally
+    	{
+    		authLock.unlock();
+    	}
+    }
+    
+    public void signalAuthComplete() {
```

Comment:
This lock processing would be much more cleanly implemented with a `CountdownLatch`.

Reply:
There's another issue here that there is no escape from this lock in the event of a disconnect.  The usual pattern is something like this:

```
while (latch.await(5, SECONDS)) {
  if (disconnected)
    break;
}
```
I wonder, though, if this would be better implemented by performing the auth negotiation _before_ returning complete in the `Completable` returned by `connect()`.  That way you can be sure that this handshaking process will definitely have completed before the caller starts asking for subscriptions.

I suspect a few other exchanges are vulnerable to this issue (particularly Bitfinex, which I'm working on at the moment), so thank you for highlighting it!

Reply:
The Completable abstract class provides blocking await functions and an overload for a timeout. Callers can wait on the connect for a specific duration before deciding that the connect failed. We could also trigger an error on the completable on auth failures, which might be better. Should I add code to invoke OnError on auth failures?